### ISSUE TITLE
Cannot use login.live.com as an OIDC authority 

### ISSUE THREAD

**Issue by hpsin** [AzureAD/microsoft-authentication-library-for-js#1824]
>> Please follow the issue template below. Failure to do so will result in a delay in answering your question.

## Library
- [x] `@azure/msal-browser@2.0.0-beta.0`

## Framework
Ad-hoc testing in Browser F12 console. 

## Description
Would like to use the login.live.com OAuth2 provider that is OIDC compliant by providing it as the authority.  However, a runtime error occurs when instantiating MSAL before any request is made. 

## Error Message
msal-browser.js:2173 Uncaught ClientConfigurationError: URL could not be parsed into appropriate segments. **Given Error: Given url string: https://login.live.com/**
     at ClientConfigurationError.AuthError [as constructor] (https://alcdn.msauth.net/browser/2.0.0-beta.0/js/msal-browser.js:191:32)
     at ClientConfigurationError.ClientAuthError [as constructor] (https://alcdn.msauth.net/browser/2.0.0-beta.0/js/msal-browser.js:416:32)
     at new ClientConfigurationError (https://alcdn.msauth.net/browser/2.0.0-beta.0/js/msal-browser.js:1473:32)
     at Function.ClientConfigurationError.createUrlParseError (https://alcdn.msauth.net/browser/2.0.0-beta.0/js/msal-browser.js:1508:20)
     at UrlString.validateAsUri (https://alcdn.msauth.net/browser/2.0.0-beta.0/js/msal-browser.js:2173:48)
     at AadAuthority.set [as canonicalAuthority] (https://alcdn.msauth.net/browser/2.0.0-beta.0/js/msal-browser.js:2570:42)
     at AadAuthority.Authority (https://alcdn.msauth.net/browser/2.0.0-beta.0/js/msal-browser.js:2554:37)
     at new AadAuthority (https://alcdn.msauth.net/browser/2.0.0-beta.0/js/msal-browser.js:2736:27)
     at Function.AuthorityFactory.createInstance (https://alcdn.msauth.net/browser/2.0.0-beta.0/js/msal-browser.js:2817:28)
     at new AuthorizationCodeModule (https://alcdn.msauth.net/browser/2.0.0-beta.0/js/msal-browser.js:2846:63)

## Security
- [ ] Is this issue security related?

## Regression
- [ ] Did this behavior work before? 
Version:

## MSAL Configuration

```js
const msalConfig = {
  auth: {
    clientId: "292666", // valid LiveID app ID. 
    authority: "https://login.live.com", //The MSA OIDC authority (e.g. https://login.live.com/.well-known/openid-configuration is off of this)
    redirectUri: "https://localhost",
  },
  cache: {
    cacheLocation: "sessionStorage", // This configures where your cache will be stored
    storeAuthStateInCookie: false, // Set this to "true" if you are having issues on IE11 or Edge
  }
};
```

## Reproduction steps
```js
const myMSALObj = new msal.PublicClientApplication(msalConfig);
```

## Expected behavior

Able to continue on to requesting an auth code. 

## Browsers/Environment
- [x] Chrome
- [ ] Firefox
- [ ] Edge
- [ ] Safari
- [ ] IE
- [ ] Other (Please add browser name here)

**Comment by jmckennon** [comment]
@hpsin Thanks for raising this, @pkanher617 will be primarily driving these 2.0 bugfixes. Moving it over to the bug board to take a deeper look.

**Comment by hpsin** [comment]
Additional details:
1. If you remove the segments check in line 2172 (  if (!components.HostNameAndPort || !components.PathSegments || components.PathSegments.length < 0) {) then the object can be created.  This line appears to assume that all authorities will have a /tenant/ section in their authority.  
1. Once the object is created, authorization endpoint lookup fails because MSAL assumes that the authority is an AAD endpoint, and uses the internal lookup API to do discovery, which fails:
     `https://login.microsoftonline.com/common/discovery/instance?api-version=1.0&authorization_endpoint=https://login.live.com/oauth2/v2.0/authorize`
1. There's a smart lookup that tries to do OIDC discovery, but only when the authority is a known OIDC endpoint:
```
if (this.isInTrustedHostList(this.canonicalAuthorityUrlComponents.HostNameAndPort)) {
                                return [2 /*return*/, this.defaultOpenIdConfigurationEndpoint];
...
    var AADTrustedHostList = [
        "login.windows.net",
        "login.chinacloudapi.cn",
        "login.cloudgovapi.us",
        "login.microsoftonline.com",
        "login.microsoftonline.de",
        "login.microsoftonline.us", 
    ];
```
I think this is B2C logic  -if you can't find the AAD authority, fall back to checking B2C.  I'd have expected the reverse, where an authority is assumed to be 3rd party OIDC instead of a special Microsoft IDP that will be in a lookup on login.mso.com.

**Comment by tnorling** [comment]
> 2. Once the object is created, authorization endpoint lookup fails because MSAL assumes that the authority is an AAD endpoint, and uses the internal lookup API to do discovery, which fails:
>    `https://login.microsoftonline.com/common/discovery/instance?api-version=1.0&authorization_endpoint=https://login.live.com/oauth2/v2.0/authorize`

@hpsin This part can be resolved by passing `knownAuthorities: ["login.live.com"]` to your `msalConfig.auth` config but it will then fail on the openid-configuration call because we currently only support the v2.0 endpoint. Supporting the v1 well-known endpoint is something we can look into for scenarios such as this.

> I think this is B2C logic -if you can't find the AAD authority, fall back to checking B2C. I'd have expected the reverse, where an authority is assumed to be 3rd party OIDC instead of a special Microsoft IDP that will be in a lookup on login.mso.com.

The logic actually is as you expect, sort of. If the previously mentioned `knownAuthorities` parameter is provided, we assume B2C/ADFS. Then if not, we fallback to the AAD lookup. Additionally while in theory 3P OIDC IDPs could work, we don't officially support that scenario.

**Comment by hpsin** [comment]
Thanks @tnorling ! That's helpful.  I'm a bit confused about what fails in the OIDC config part - .well-known/... is not a v2-specific item, so should work on all platforms. Note this isn't the v1 well-known endpoint, it's just "a well known endpoint off of an OIDC authority"

That said - ACK, thanks for confirming.  It would be great to see MSAL be used for other providers but I understand why it's not.

**Comment by tnorling** [comment]
@hpsin To clarify, we currently build the well-known endpoint like so: `authority + "/v2.0/.well-known/openid-configuration"` unless the authority is an adfs authority in which case we leave off the "v2.0" portion of the url

**Comment by hpsin** [comment]
Ahhh, right, That's a bummer :( I'd have assumed that people use the https://login.microsoftonline.com/common/v2.0/ as the authority.  Makes sense.

**Comment by DarylThayil** [comment]
@hpsin  does @tnorling  suggestion using known authorities to skip openId config work for you

**Comment by hpsin** [comment]
No, because it falls back to calling :
https://login.microsoftonline.com/common/discovery/instance?api-version=1.0&authorization_endpoint=https://login.live.com/oauth2/v2.0/authorize 400 (Bad Request)

That's with 
```
var msalConfig = {
  auth: {
    clientId: "292666",
    validateAuthority:false,
    knownAuthorities: ["login.live.com"],
    authority: "https://login.live.com/",
    redirectUri: "https://localhost",
  },
```

**Comment by tnorling** [comment]
Hmm I would not expect it to hit that endpoint when providing `knownAuthorities`. I just copied that config to try it myself and it's failing on the openId-config call as I would expect because it's trying to hit `https://login.live.com/v2.0/.well-known/openid-configuration` instead of `https://login.live.com/.well-known/openid-configuration`. 

In any event, I don't expect this scenario to work without additional changes to the library so we'll need to discuss with the team if this should be a supported scenario and, if so, what those changes should look like. 

To summarize there's 2 issues currently preventing the library from using `login.live.com` as an authority:
1. Library expects tenant to always be a part of the provided authority
2. Library always appends "v2.0" to the authority when making the `.well-known/openid-configuration` call

**Comment by jasonnutter** [comment]
@jmckennon will address this alongside #2120

**Comment by tom171296** [comment]
Are there any update or workarounds for this issue?

**Comment by jmckennon** [comment]
@tom171296 I'm currently working on this fix, I will link the PR when it's up.

**Comment by hpsin** [comment]
🎉🎉🎉🎉🎊🎊🎊🎊🥳🥳🥳🥳
This is awesome. I can't wait to try it out! This is the first officially recommendable library for login.live.com in a while.
